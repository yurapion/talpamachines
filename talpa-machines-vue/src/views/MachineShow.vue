<template>
  <v-container>
    <div class="machine-header">
      <h1 class="title">This is a {{ machine.name }}</h1>
    </div>

    <h4 v-if="!checkSensor">Please Select a Sensor</h4>

    <v-select
      @change="onChange($event)"
      :items="machine.sensors"
      v-model="sensor"
      item-text="name"
      item-value="id"
      label="Sensor"
      solo
    ></v-select>

    <div class="location">
      <!-- <BaseIcon name="map" class="icon"><h2>Location</h2></BaseIcon> -->

      <address>Last known gps position {{ machine.lastKnownPosition }}</address>
    </div>

    <form
      v-if="checkSensor"
      @submit.prevent="fetchSesordata(id, startDate, endDate, sensor)"
    >
      <!-- <input type="text" name="startDate" v-model="startDate" />
      <input type="text" name="endDate" v-model="endDate" /> -->
      <!-- <div v-if="checkSensor"> -->
      <div>
        <h3 v-if="!checkDates">Please Select Dates</h3>
        <div class="field">
          <label>First day</label>
          <datepicker
            v-model="startDate"
            placeholder="Select a start date"
            class="date"
          />
        </div>

        <div class="field">
          <label>Last day</label>
          <datepicker
            v-model="endDate"
            placeholder="Select an end date"
            class="date"
          />
        </div>
      </div>

      <v-btn type="submit" v-if="checkDates" color="info"
        >Fetch Sensor Data
      </v-btn>
    </form>

    <DataGraph v-if="checkIfData" cols="12" md="4" :data="chartSensorsData" />

    <!-- <BaseIcon v-if="!chechDataAfterSubmit" name="info">
      Sorry there are no data for selected dates for {{ sensor.name }}</BaseIcon
    > -->
    <v-snackbar v-model="snackbar" :left="$vuetify.breakpoint.lgAndUp">
      You have selected {{ selectedMachine.name }}
      <v-btn color="pink" text @click="snackbar = false">
        Close
      </v-btn>
    </v-snackbar>
  </v-container>
</template>

<script>
import Datepicker from "vuejs-datepicker";
import { mapState, mapGetters } from "vuex";
import DataGraph from "../components/DataGraph";
export default {
  components: {
    Datepicker,
    DataGraph
  },
  props: ["id"],
  data() {
    return {
      startDate: "",
      endDate: "",
      sensor: {},
      change: true,
      selectedMachine: {
        name: "",
        id: ""
      },
      snackbar: false
    };
  },
  computed: {
    checkDates() {
      return this.startDate && this.endDate;
    },
    checkSensor() {
      return this.sensor > 0;
    },
    chechDataAfterSubmit() {
      return this.change || this.checkIfData;
    },
    ...mapGetters(["chartSensorsData", "checkIfData"]),
    ...mapState(["machine", "sensorDataList"])
  },
  methods: {
    fetchSesordata(id, startDate, endDate, sensorId) {
      this.$store.dispatch("fetchSensorDataList", {
        id,
        sensorId,
        startDate,
        endDate
      });
      this.change = false;
    },
    onChange() {
      this.change = true;
    }
  },
  beforeCreate() {},
  created() {
    this.$store.dispatch("fetchMachine", this.id);

    this.snackbar = true;
    this.selectedMachine.name = this.machine.name;
    this.selectedMachine.id = this.machine.id;
  },
  mounted() {}
};
</script>

<style scoped>
.date {
  width: auto;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.2);
}
.location {
  margin-bottom: 0;
}
.location > .icon {
  margin-left: 10px;
}
.machine-header > .title {
  margin: 0;
}
.field {
  margin-bottom: 24px;
  width: 20%;
}
</style>
