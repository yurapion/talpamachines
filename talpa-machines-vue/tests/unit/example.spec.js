import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import MachineShow from "../../src/views/MachineShow.vue";
import Vuetify from "vuetify";
import { makeExecutableSchema, addMockFunctionsToSchema } from "graphql-tools";
import { graphql } from "graphql";

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Vuetify);

describe("MachineShow.vue", () => {
  let actions;
  let store;
  let state;
  let getters;
  let obj = {
    a: 1
  };
  let vuetify;

  beforeEach(() => {
    state = {
      machine: { name: "Buldoser", id: 1, sensors: [{ name: "Speed", id: 1 }] }
    };
    actions = {
      fetchSensorDataList: jest.fn(),
      fetchMachine: jest.fn()
    };
    getters = {
      checkIfData: () => true,
      chartSensorsData: () => obj
    };
    store = new Vuex.Store({
      state,
      actions,
      getters
    });
    vuetify = new Vuetify();
  });
  it("call action fetchSensorDataList when form submit", () => {
    const wrapper = shallowMount(MachineShow, {
      computed: { checkSensor: () => true },
      // mocks: {
      //   $vuetify: { breakpoint: {} }
      // },
      store,
      localVue,
      vuetify
    });
    // wrapper.setData({ sensor: { a: 1 } });
    // wrapper.setData({ checkSensor: () => true });
    wrapper.find("form").trigger("submit");
    expect(actions.fetchSensorDataList).toHaveBeenCalled();
  });

  it("call action fetchMachine  when created", () => {
    const wrapper = shallowMount(MachineShow, {
      computed: { checkSensor: () => true },
      store,
      localVue,
      vuetify
    });

    wrapper.find(".title");
    expect(actions.fetchMachine).toHaveBeenCalled();
  });

  it("check if snaphot is correct", () => {
    const wrapper = shallowMount(MachineShow, {
      computed: { checkSensor: () => true },
      store,
      localVue,
      vuetify
    });

    // expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find(".title").text()).toBe("This is a Buldoser");
  });

  //////////////////////////////////////// SHOULD ADD APPOLLO DIRECTLY TO THE COMPONENT.
  /////////////////////////////////////// TEST DOESN'T WORK BECAUSE ALL QUERIES ARE GOING THROUGH VUEX.
  it("Query data GraphQl", () => {
    const wrapper = shallowMount(MachineShow, {
      computed: { checkSensor: () => true },
      mocks: { $appollo: { queries: { machine: {} } } },
      store,
      localVue,
      vuetify
    });

    const sourceSchema = `
  type Machine {
    id: Int!
    name: String!
  }

  type Query {
    machines: [Machine]
  }
`;

    const query = `
  query {
    machines {
      id
      name
    }
  }
`;

    const schema = makeExecutableSchema({
      typeDefs: sourceSchema
    });

    addMockFunctionsToSchema({
      schema
    });

    graphql(schema, query).then(result => {
      wrapper.setData(result.data);
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  // it("should emit an event when the action v-select is changed", () => {
  //   const wrapper = shallowMount(MachineShow, {
  //     computed: { checkSensor: () => true },
  //     store,
  //     localVue,
  //     vuetify
  //   });

  //   const event = jest.fn();
  //   const select = wrapper.find(".v-select");

  //   wrapper.vm.$on("@change", event);

  //   expect(event).toHaveBeenCalledTimes(0);
  //   console.log(wrapper.vm.machine.sensors);
  //   select.trigger("change");

  //   expect(event).toHaveBeenCalledTimes(1);
  // });
});
