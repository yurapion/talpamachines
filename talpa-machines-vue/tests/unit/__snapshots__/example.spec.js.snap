// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MachineShow.vue Query data GraphQl 1`] = `
<v-container-stub
  tag="div"
>
  <div
    class="machine-header"
  >
    <h1
      class="title"
    >
      This is a Buldoser
    </h1>
  </div>
   
  <!---->
   
  <v-select-stub
    appendicon="$dropdown"
    backgroundcolor=""
    clearicon="$clear"
    errorcount="1"
    errormessages=""
    itemcolor="primary"
    itemdisabled="disabled"
    items="[object Object]"
    itemtext="name"
    itemvalue="id"
    label="Sensor"
    loaderheight="2"
    menuprops="[object Object]"
    messages=""
    nodatatext="$vuetify.noDataText"
    rules=""
    solo="true"
    successmessages=""
    type="text"
    value="[object Object]"
    valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
  />
   
  <div
    class="location"
  >
    <address>
      Last known gps position 
    </address>
  </div>
   
  <form>
    <div>
      <h3>
        Please Select Dates
      </h3>
       
      <div
        class="field"
      >
        <label>
          First day
        </label>
         
        <datepicker-stub
          class="date"
          format="dd MMM yyyy"
          language="[object Object]"
          maximumview="year"
          minimumview="day"
          placeholder="Select a start date"
          value=""
        />
      </div>
       
      <div
        class="field"
      >
        <label>
          Last day
        </label>
         
        <datepicker-stub
          class="date"
          format="dd MMM yyyy"
          language="[object Object]"
          maximumview="year"
          minimumview="day"
          placeholder="Select an end date"
          value=""
        />
      </div>
    </div>
     
    <!---->
  </form>
   
  <datagraph-stub
    cols="12"
    data="[object Object]"
    md="4"
  />
   
  <v-snackbar-stub
    timeout="6000"
    value="true"
  >
    
    You have selected Buldoser
    
    <v-btn-stub
      activeclass=""
      color="pink"
      tag="button"
      text="true"
      type="button"
    >
      
      Close
    
    </v-btn-stub>
  </v-snackbar-stub>
</v-container-stub>
`;

exports[`MachineShow.vue check if snaphot is correct 1`] = `
<v-container-stub tag="div">
  <div class="machine-header">
    <h1 class="title">This is a Buldoser</h1>
  </div>
  <!---->
  <v-select-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="[object Object]" appendicon="$dropdown" backgroundcolor="" label="Sensor" loaderheight="2" clearicon="$clear" solo="true" type="text" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" nodatatext="$vuetify.noDataText" items="[object Object]" itemcolor="primary" itemdisabled="disabled" itemtext="name" itemvalue="id" menuprops="[object Object]"></v-select-stub>
  <div class="location">
    <address>Last known gps position </address>
  </div>
  <form>
    <div>
      <h3>Please Select Dates</h3>
      <div class="field"><label>First day</label>
        <datepicker-stub value="" format="dd MMM yyyy" language="[object Object]" placeholder="Select a start date" minimumview="day" maximumview="year" class="date"></datepicker-stub>
      </div>
      <div class="field"><label>Last day</label>
        <datepicker-stub value="" format="dd MMM yyyy" language="[object Object]" placeholder="Select an end date" minimumview="day" maximumview="year" class="date"></datepicker-stub>
      </div>
    </div>
    <!---->
  </form>
  <datagraph-stub data="[object Object]" cols="12" md="4"></datagraph-stub>
  <v-snackbar-stub value="true" timeout="6000">
    You have selected Buldoser
    <v-btn-stub color="pink" tag="button" activeclass="" text="true" type="button">
      Close
    </v-btn-stub>
  </v-snackbar-stub>
</v-container-stub>
`;
